rust/ql/lib/codeql/rust/frameworks/stdlib/lang-core.model.yml

```diff
@@ -40,7 +40,11 @@ extensions:
       - ["lang:core", "crate::ptr::write_unaligned", "Argument[1]", "Argument[0].Reference", "value", "manual"]
       - ["lang:core", "crate::ptr::write_volatile", "Argument[1]", "Argument[0].Reference", "value", "manual"]
       # Str
+      - ["lang:core", "<str>::as_str", "Argument[self]", "ReturnValue", "taint", "manual"]
+      - ["lang:core", "<str>::as_bytes", "Argument[self]", "ReturnValue", "taint", "manual"]
+      - ["lang:core", "<str>::to_string", "Argument[self]", "ReturnValue", "taint", "manual"]
```

Comment:
The function `<str>::as_str` is value preserving, it just returns `self`. I think `as_bytes` should be considered `value` preserving as well. It is defined as and doesn't change the value, only its type.
```
    pub const fn as_bytes(&self) -> &[u8] {
        // SAFETY: const sound because we transmute two types with the same layout
        unsafe { mem::transmute(self) }
    }
```

I'm a little in doubt of whether to consider `str::to_string` to be value preserving as well. I'm not sure whether we consider copies to be taint or value preserving in other cases.

Reply:
Good idea.  Done.

> I'm a little in doubt of whether to consider str::to_string to be value preserving as well. 

I would say no, because *in general* `to_string` is not value preserving.